File: manual/data/cpp/tests/test_catboost_importance.cpp
#include <gtest/gtest.h>
#include "../src/function_catboost_importance.cpp"

class TestBoostFeatureImportance : public ::testing::Test {
protected:
    void SetUp() override {
        importance_calculator = new BoostFeatureImportance();
    }

    void TearDown() override {
        delete importance_calculator;
    }

    BoostFeatureImportance* importance_calculator;
};

TEST_F(TestBoostFeatureImportance, TestPredictionValuesChange) {
    importance_calculator->process("train", 1.0f, 2.0f, 3.0f, 0);
    float result = importance_calculator->process("importance", 1.0f, 2.0f, 3.0f, -1, "PredictionValuesChange");
    ASSERT_GE(result, 0.0f);
}

TEST_F(TestBoostFeatureImportance, TestLossFunctionChange) {
    importance_calculator->process("train", 4.0f, 5.0f, 6.0f, 1);
    float result = importance_calculator->process("importance", 4.0f, 5.0f, 6.0f, -1, "LossFunctionChange");
    ASSERT_GE(result, 0.0f);
}

TEST_F(TestBoostFeatureImportance, TestShapValues) {
    importance_calculator->process("train", 7.0f, 8.0f, 9.0f, 0);
    float result = importance_calculator->process("importance", 7.0f, 8.0f, 9.0f, -1, "ShapValues");
    ASSERT_GE(result, 0.0f);
}

TEST_F(TestBoostFeatureImportance, TestInvalidImportanceType) {
    importance_calculator->process("train", 10.0f, 11.0f, 12.0f, 1);
    ASSERT_THROW(
        importance_calculator->process("importance", 10.0f, 11.0f, 12.0f, -1, "InvalidType"),
        std::invalid_argument
    );
}

TEST_F(TestBoostFeatureImportance, TestDifferentInputs) {
    importance_calculator->process("train", 1.5f, 2.5f, 3.5f, 0);
    float result_1 = importance_calculator->process("importance", 1.5f, 2.5f, 3.5f, -1, "PredictionValuesChange");
    importance_calculator->process("train", 4.5f, 5.5f, 6.5f, 1);
    float result_2 = importance_calculator->process("importance", 4.5f, 5.5f, 6.5f, -1, "PredictionValuesChange");
    ASSERT_NE(result_1, result_2);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: manual/data/python/tests/test_catboost_importance.py
import unittest
from function_catboost_importance import CatBoostFeatureImportance

class TestCatBoostFeatureImportance(unittest.TestCase):
    def setUp(self):
        self.importance_calculator = CatBoostFeatureImportance()

    def test_prediction_values_change(self):
        self.importance_calculator.process("train", 1.0, 2.0, 3.0, 0)
        result = self.importance_calculator.process("importance", 1.0, 2.0, 3.0, importance_type="PredictionValuesChange")
        self.assertIsInstance(result, float)

    def test_loss_function_change(self):
        self.importance_calculator.process("train", 4.0, 5.0, 6.0, 1)
        result = self.importance_calculator.process("importance", 4.0, 5.0, 6.0, importance_type="LossFunctionChange")
        self.assertIsInstance(result, float)

    def test_shap_values(self):
        self.importance_calculator.process("train", 7.0, 8.0, 9.0, 0)
        result = self.importance_calculator.process("importance", 7.0, 8.0, 9.0, importance_type="ShapValues")
        self.assertIsInstance(result, float)

    def test_invalid_importance_type(self):
        self.importance_calculator.process("train", 10.0, 11.0, 12.0, 1)
        with self.assertRaises(ValueError):
            self.importance_calculator.process("importance", 10.0, 11.0, 12.0, importance_type="InvalidType")

    def test_different_inputs(self):
        self.importance_calculator.process("train", 1.5, 2.5, 3.5, 0)
        result_1 = self.importance_calculator.process("importance", 1.5, 2.5, 3.5, importance_type="PredictionValuesChange")
        self.importance_calculator.process("train", 4.5, 5.5, 6.5, 1)
        result_2 = self.importance_calculator.process("importance", 4.5, 5.5, 6.5, importance_type="PredictionValuesChange")
        self.assertNotEqual(result_1, result_2)


if __name__ == '__main__':
    unittest.main()
